[openssl 3.x.x] fix "official" OpenSSL vars for CMakeDeps gen

jngrb opened the pull request:
this extends the improvements of PR #12838 done for the 1.x.x recipe to the 3.x.x one; also there is a new 3.x.x test_package using the CMakeDeps generator

---

Reason: without this change, it is currently not possible to override the OpenSSL version to 3.0.x in Poco (see comment below)

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

uilianries commented on recipes/openssl/3.x.x/test_package/conanfile.py:
```diff
@@ -1,40 +1,67 @@
+    test_type = "explicit"
+
+    @property
+    def _skip_test(self):
```
You can't skip a test. If something does not work, please, raise a ConanInvalidConfiguration in `validate()` method in the recipe

jngrb replied:
I just copied & pasted this part from the "v2 test package" for 1.x.x: https://github.com/conan-io/conan-center-index/blame/master/recipes/openssl/1.x.x/test_package/conanfile.py#L16

I originates here: https://github.com/conan-io/conan-center-index/pull/9208#issuecomment-1027081580

As this test skipping was not part of the "v1 test package" for 3.x.x, I will give it a try not to have this in the "v2 test package" for 3.x.x

jngrb commented:
Giving some background for this change: I would like to be able to override openssl to version 3.0.7 in Poco. As of now, since Poco was changed to CMakeDeps in https://github.com/conan-io/conan-center-index/pull/12868 , this no longer works:

```
conan create -o poco:enable_data=False -o poco:enable_data_mysql=False -o poco:enable_data_postgresql=False -o poco:enable_data_sqlite=False -o poco:enable_activerecord=False -o poco:enable_redis=False -o poco:enable_mongodb=False recipes/poco/all poco/1.12.4@
````

=> works

```
conan create -o poco:enable_data=False -o poco:enable_data_mysql=False -o poco:enable_data_postgresql=False -o poco:enable_data_sqlite=False -o poco:enable_activerecord=False -o poco:enable_redis=False -o poco:enable_mongodb=False --require-override=openssl/3.0.7@ recipes/poco/all poco/1.12.4@
```
(I added a **--require-override=openssl/3.0.7@** here.)

=> fails as of now, because Poco's CMakeLists.txt does not see a positive OPENSSL_FOUND and thus I run into #13577 (so far only fixed for OpenSSL 1.x.x)
=> will be fixed once this PR was accepted

(Note: I only disabled any irrelevant Poco component just to have it compile faster so that I get results more quickly.)

jngrb commented:
Based on my tests with the last commits, I see that the MacOS failures are not introduced with this PR. It seems that the legacy provider from the shared object `ossl-modules/legacy.so` does not work on MacOS even before this PR. I cannot find a solution because I don't have an Apple device to build on.

A workaround is to either use `openssl:no_legacy = True` so that there are no legacy algorithms at all. Or to use `openssl:no_module = True` to compile the legacy providers directly into the main library.

As I cannot change the default value for these options without introducing breaking changes for the consumer, I can only raise a ConanInvalidConfiguration exception (in line with the [comment](https://github.com/conan-io/conan-center-index/pull/14426#discussion_r1032605657) above) for `os = MacOS` and `openssl:no_legacy = False && openssl:no_module = False` until this is fixed.

prince-chrismc commented:
#15802 has the same problem 😖

jngrb commented:
Now that !15802 was merged, I rebased my changes to the current master. I also reverted the changes from this comment of mine:

> Based on my tests with the last commits, I see that the MacOS failures are not introduced with this PR. It seems that the legacy provider from the shared object `ossl-modules/legacy.so` does not work on MacOS even before this PR. I cannot find a solution because I don't have an Apple device to build on.
> 
> A workaround is to either use `openssl:no_legacy = True` so that there are no legacy algorithms at all. Or to use `openssl:no_module = True` to compile the legacy providers directly into the main library.
> 
> As I cannot change the default value for these options without introducing breaking changes for the consumer, I can only raise a ConanInvalidConfiguration exception (in line with the [comment](https://github.com/conan-io/conan-center-index/pull/14426#discussion_r1032605657) above) for `os = MacOS` and `openssl:no_legacy = False && openssl:no_module = False` until this is fixed.

compare with this comment in !15802:

https://github.com/conan-io/conan-center-index/pull/15802#issuecomment-1431750930

Croydon commented on recipes/openssl/3.x.x/test_package/conanfile.py:
```diff
@@ -1,40 +1,54 @@
+        if self.settings.os == "Android":
+            tc.cache_variables["CONAN_LIBCXX"] = ""
+        openssl = self.dependencies["openssl"]
+        openssl_version = Version(openssl.ref.version)
```
```suggestion
```

Unused

Croydon commented on recipes/openssl/3.x.x/test_v1_package/CMakeLists.txt:
```diff
@@ -0,0 +1,55 @@
+# are properly defined in conan generators
+set(_custom_vars
+    OPENSSL_FOUND
+    OPENSSL_INCLUDE_DIR
+    OPENSSL_CRYPTO_LIBRARY
+    OPENSSL_CRYPTO_LIBRARIES
+    OPENSSL_SSL_LIBRARY
+    OPENSSL_SSL_LIBRARIES
+    OPENSSL_LIBRARIES
+    OPENSSL_VERSION
+)
+foreach(_custom_var ${_custom_vars})
+    if(DEFINED ${_custom_var} AND NOT "${${_custom_var}}" STREQUAL "" )
+        message(STATUS "${_custom_var}: ${${_custom_var}}")
+    else()
+        message(FATAL_ERROR "${_custom_var} not defined")
+    endif()
+endforeach()
+
+add_executable(digest ../test_package/digest.c)
+if(OPENSSL_WITH_ZLIB)
+    target_compile_definitions(digest PRIVATE WITH_ZLIB)
+endif()
+target_link_libraries(digest OpenSSL::Crypto)
+
+if(OPENSSL_WITH_LEGACY)
+    add_executable(digest_legacy ../test_package/digest_legacy.c)
+    # do now show deperecation warnings
+    target_compile_definitions(digest_legacy PRIVATE OPENSSL_SUPPRESS_DEPRECATED)
+    if(OPENSSL_WITH_MD4)
+      target_compile_definitions(digest_legacy PRIVATE OPENSSL_WITH_MD4)
+    endif()
+    if(OPENSSL_WITH_RIPEMD160)
+      target_compile_definitions(digest_legacy PRIVATE OPENSSL_WITH_RIPEMD160)
+    endif()
+    if(OPENSSL_WITH_ZLIB)
+        target_compile_definitions(digest_legacy PRIVATE WITH_ZLIB)
+    endif()
+    target_link_libraries(digest_legacy OpenSSL::Crypto)
+endif()
```
```suggestion
cmake_minimum_required(VERSION 3.1)
project(test_package C)

include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
conan_basic_setup(TARGETS)

add_subdirectory(${CMAKE_CURRENT_SOURCE_DIR}/../test_package/
                 ${CMAKE_CURRENT_BINARY_DIR}/test_package/)
```

Re-use the CMake + C source from `test_package` as much as possible

Croydon commented on recipes/openssl/3.x.x/test_v1_package/conanfile.py:
```diff
@@ -0,0 +1,40 @@
+        cmake.build()
+
+    def test(self):
+        if not cross_building(self):
```
```suggestion
        if not tools.cross_building(self):
```

Croydon commented on recipes/openssl/3.x.x/test_v1_package/conanfile.py:
```diff
@@ -0,0 +1,40 @@
+from conans import CMake, tools, ConanFile
+from conan.tools.build import cross_building
```
```suggestion
```

Croydon commented:
Please only do force-pushes when really necessary, as it causes reviewers to start all over again going through the changes.

Also please use either the GitHub review function to accept code changes or add the review as a commit co-author in your git client to give some credits for their work ✌️

Croydon commented on recipes/openssl/3.x.x/test_package/conanfile.py:
```diff
@@ -1,40 +1,53 @@
-from conans import CMake, tools, ConanFile
-from conan.tools.build import cross_building
+from conan import ConanFile
+from conan.tools.scm import Version
```
```suggestion
```

Follow up error, unused

Croydon commented on recipes/openssl/3.x.x/test_package/conanfile.py:
```diff
@@ -1,40 +1,53 @@
                 self.run(bin_legacy_path, run_environment=True)
-
-            if not self.options["openssl"].no_stdio:
-                self.run("openssl version", run_environment=True)
```
Why did you delete this?

jngrb replied:
Will be readded

Croydon commented on recipes/openssl/3.x.x/test_package/conanfile.py:
```diff
@@ -1,40 +1,53 @@
         assert os.path.exists(os.path.join(self.deps_cpp_info["openssl"].rootpath, "licenses", "LICENSE.txt"))
-
-        for fn in ("libcrypto.pc", "libssl.pc", "openssl.pc",):
-            assert os.path.isfile(os.path.join(self.build_folder, fn))
```
Why did you delete this?

jngrb replied:
because these files are no lot generated for the CMakeDeps generator.

 I would need the cmake, cmake_find_package, and pkg_config generators as for the test_v1_package test package (where this check is included).

Also compare that the *.pc files are also only checked for the 1.x.x test_v1_package, not for the 1.x.x test_package.

jngrb commented:
Attempted the conan v2 port for OpenSSL 3.x.x in PR #16563

roalter commented:
Is there a reason why this is still in the queue?

jcar87 commented:
> Is there a reason why this is still in the queue? @prince-chrismc @jcar87 : Maybe some of you _could_ give it a try to review & integrate? There are still getting CVEs on OpenSSL; picking newer OpenSSL versions a little bit faster would be a great opportunity.

Hi @roalter - this is on my stack today, including for support for Conan 2.0 for this recipe. The issues that @jngrb  mentions in https://github.com/conan-io/conan-center-index/pull/14426#issuecomment-1327925221 were addressed not long ago.
On the other hand, if I'm not mistaken the issue addressed in this PR is not a blocker for adding new versions.

I will make sure to include version 3.1.0 as well, thanks for the heads up!

jcar87 commented:
Hi @jngrb - I've partially added the changes from this PR into this one, preserving your authorship:
https://github.com/conan-io/conan-center-index/pull/16658

Please let me know if this addresses the issues with poco that you have mentioned. If there are any other changes from this PR that you wish you bring forward, please let us know once the PR branch has been sync'ed up with the head of master, and we'll happily review this. If the recipe is already satisfactory, please close this PR :) 

Thanks for your contribution!

jngrb commented:
> Hi @jngrb - I've partially added the changes from this PR into this one, preserving your authorship: #16658
> 
> Please let me know if this addresses the issues with poco that you have mentioned. If there are any other changes from this PR that you wish you bring forward, please let us know once the PR branch has been sync'ed up with the head of master, and we'll happily review this. If the recipe is already satisfactory, please close this PR :)

Thank you. All my requirements are now met. I close this PR.
